src/ui/toolbox/menu.js

```diff
@@ -117,6 +117,30 @@ var menuComponentApi = {
         }
     },
 
+    /**
+     * It needs to find nearest visible item.
+     *
+     * @param {-1|1} inc - incrementor. -1 - navigate to the top,  1 - to the bottom.
+     *
+     * @returns {JQuery} element.
+     */
+    _findClosestVisibleItem(inc = 1) {
+        if (!this.menuItems.length) {
+            return;
+        }
+        const last = inc === 1 ? this.menuItems.length : -1;
+        let elem;
+        do {
+            this.hoverIndex += inc;
+            if (this.hoverIndex === last) {
+                return;
+            }
+            elem = this.menuItems[this.hoverIndex].getElement();
+        } while (elem && elem.hasClass('hidden'));
```

Comment:
Why not rely on `lodash` for such an algorithm?

You migt have a clearer code using `_.find()` and `_.findLast()`.

Also you might simplify it by having 2 methods instead: `findFirstVisibleItem()` and `findLastVisibleItem()`. This will make the code more readable.

Reply:
Lodash methods are not fit, because we  rely on the hover index (it can be for instance 5). Also this function have to change hoverIndex, because menuItems can contain invisible items and we should skip them. Nonetheless I'll rename it to improve DX. 

Good point about 2 methods, I'll use currying

Reply:
But this is exactly one of the flaws of the function: it has a side effect.
A better implementation would be to only compute the index, without modifying it.
Then the consumer that will call this function will have to update actually the index.

Reply:
Most of the methods of this object are not pure, so I'm not sure that I can understand, whats wrong just with this one. Developer can easily understand by the name of the function that it applies some necessary side effects.

Anyway of course it's possible to split function into two: **hoverClosestVisibleItem** and **findClosestVisibleItem** the last one will be called by the first one. Is it ok?

Reply:
Yes, unfortunately, the existing code might not follow the best practices. However, the goal is to avoid continuing the mess. So if the fix requires some refactoring, don't hesitate to do so.